src/ctypes-foreign-threaded/foreign.mli

```diff
@@ -93,3 +104,88 @@ val funptr_opt :
+
+  type t
+  (** Handle to an ocaml function that can be passed to C for use in callbacks.
+
+      Unfortunately it is not possible to track if a fuction pointer is still used in C,
+      so you must use the appropriate life cycle functions below for this.
+
+      See [t], [of_fun], [with_fun] and [free] *)
+
+  val t : t Ctypes.typ
+  (** ctype that can be used in type signatures of external functions *)
+
+  val t_opt : t option Ctypes.typ
+  (** Like [t] but optional. *)
+
+  val free : t -> unit
+  (** [free fptr] - Indicate that the [fptr] is no longer needed.
+
+      Once [free] has been called any C calls to this [Dynamic_funptr.t] are
+      unsafe and may result in a segmentation fault. Only call [free] once the
+      callback is no longer used from C.
+  *)
+
+  val of_fun : ?debug_info:string -> fn -> t
+  (** [of_fun fn] - Turn an ocaml closure into a function pointer
+      that can be passed to C.
+
+      You MUST call [free] to the function pointer is no longer needed.
+      Failure to do so will result in a memory leak.
+      Failure to call [free] and not holding a reference this this pointer
+      is an error. To avoid unexpected crashes we log this and ensure that
+      potentially still needed ocaml values are retained.
+
+      For many use cases it may be simpler to use [with_fun] instead as this will
+      do the free for you. *)
+
+  val with_fun : ?debug_info:string -> fn -> (t -> 'c) -> 'c
+  (** [with_fun fn (fun fptr -> DO_STUFF)] - Turn an ocaml closure into a
+      function pointer and do simple life cycle management.
+
+      This will automatically call [free fptr] after [DO_STUFF] completes.
+
+      It is not safe to use if the C function ptr [fptr] may still be used
+      after [DO_STUFF] completed.
+  *)
+end
+
+(** [(val (dynamic_funptr fn))] - define a Ctype for more safely passing ocaml
+    functions to C.
+
+    [(val (funptr_spec (FOO @-> returning BAR)))] is roughly equivalent to
+    [BAR( * )(FOO)] in C.
+
+    Example:
+    {[
+      module Progress_callback = (val (dynamic_funptr (int @-> int @-> ptr void @-> returning void)))
+      let keygen =
+        foreign "RSA_generate_key" (int @-> int @-> Progress_callback.t @-> ptr void @-> returning rsa_key)
+      let secret_key =
+        Progress_callback.with_fun
+          (fun a b _ -> printf "progress: a:%d, b:%d\n" a b)
+          (fun progress ->
+             keygen 2048 65537 progress null)
+    ]}
+*)
+val dynamic_funptr
+  :  ?abi:Libffi_abi.abi
+  -> ?runtime_lock:bool
+  -> ?thread_registration:bool
+  -> ('a -> 'b) Ctypes.fn
+  -> (module Funptr with type fn = 'a->'b)
+
+val call_static_funptr :
+  ?name:string ->
+  ?abi:Libffi_abi.abi ->
+  ?check_errno:bool ->
+  ?release_runtime_lock:bool ->
+  ('a -> 'b) Ctypes.fn ->
+  ('a -> 'b) Ctypes.static_funptr -> 'a -> 'b
+(** [call_static_funptr fn fptr] - Helper to allow calling function pointers passed to
+    ocaml from C. *)
```

Comment:
Let's leave `call_static_funptr` out of the interface:
(1) it should be possible to achieve the same thing via `Ctypes.coerce`
(2) `call_static_funptr` merges together coercion (which is comparatively expensive) and invocation, but it's often good practice to separate them (rather than allocating a fresh closure per call)

Reply:
This has been removed.